src/algorithms.jl

```
@@ -908,3 +908,83 @@ function pyramid_scale(img, downsample)
     sz_next = map(s->ceil(Int, s/downsample), size(img))
     imresize(img, sz_next)
 end
+
+"""
+```
+thres = otsu_threshold(img)
+thres = otsu_threshold(img, bins)
+```
+
+Computes treshold for grayscale image using Otsu's method.  
+  
+Parameters:  
+-    img         = Grayscale input image    
+-    bins        = Number of bins used to compute the histogram. Needed for floating-point images.    
+    
+"""
+function otsu_threshold{T<:FixedPointNumbers.Normed}(img::AbstractArray{Gray{T}, 2})
+    bins::Int = 2^T.parameters[2]
+    histogram = zeros(Float64, bins)
+    for v in img
+        histogram[gray(v).i + 1] += 1
+    end
+    histogram = histogram./sum(histogram)
+
+    ω0 = 0
+    μ0 = 0
+    μt = 0
+    μT = sum((1:bins).*histogram)
+    max_σb=0.0
+    thres=1
+
+    for t in 1:bins-1
+        ω0 += histogram[t]
+        ω1 = 1 - ω0
+        μt += t*histogram[t]
+
+        σb = (μT*ω0-μt)^2/(ω0*ω1)
+
+        if(σb > max_σb)
+            max_σb = σb
+            thres = t
+        end
+    end
+
+    return T((thres-1)/bins)
+end
+
+function otsu_threshold{T<:AbstractFloat}(img::AbstractArray{Gray{T}, 2}, bins::Integer)
+    
+    histogram = zeros(Float64, bins)
+    for v in img
+        index::Int = round(gray(v)*bins + 1)
+        if index == bins + 1
+            index = bins
+        end
```

What if there are values outside the range of 0-1? This currently should throw an error, which might be best. Alternatively, would it be worth checking `extrema`? The cost, of course, is two passes through the image which I can understand you might want to avoid.